atef/check.py

```diff
@@ -83,6 +90,105 @@ def compare(self, value: PrimitiveType) -> bool:
         return value == self.value
 
 
+@dataclass
+@serialization.as_tagged_union
+class DynamicValue:
+    """
+    A primitive value from an external source that may change over time.
+    This necessarily picks up a runtime performance cost and getting
+    the value is not guaranteed to be successful. If unsuccessful,
+    This will raise a DynamicValueError from the original exception.
+
+    Includes settings for reduction of multiple samples.
+
+    Value will be cached on preparation, and this value used for comparisons
+    """
+    #: Value is now optional, and will be filled in when prepared
+    value: Optional[PrimitiveType] = None
+
+    #: Period over which the value will be read
+    reduce_period: Optional[Number] = None
+
+    #: Reduce collected samples by this reduce method
+    reduce_method: reduce.ReduceMethod = reduce.ReduceMethod.average
+
+    #: If applicable, request and compare string values rather than the default
+    string: Optional[bool] = None
+
+    def __str__(self) -> str:
+        kwds = (f"{key}={value}" for key, value in asdict(self).items())
+        return f"{type(self).__name__}({', '.join(kwds)}) [{self.value}]"
+
+    def get(self) -> PrimitiveType:
+        """
+        Call the child's get routine and wrap for error handling.
+        This also sets up a cached last value to use in the repr.
+        """
+        if self.value is not None:
+            return self.value
+        else:
+            raise DynamicValueError('Dynamic value has not been prepared.')
+
+    async def prepare(self, cache: DataCache) -> None:
+        """
+        Implement in child class to get the current value from source.
+        Should set the self.value
+        """
+        raise NotImplementedError()
+
+
+@dataclass
+class EpicsValue(DynamicValue):
+    """
+    A primitive value sourced from an EPICS PV.
+    This will create and cache an EpicsSignalRO object, and defer
+    to that signal's get handling.
+    """
+    # as of 3.10, use kw_only=True to allow mandatory arguments after the inherited
+    # optional ones.  Until then, these must have a default
+    #: The EPICS PV to use.
+    pvname: str = ''
+
+    async def prepare(self, cache: DataCache) -> None:
+        if not self.pvname:
+            raise DynamicValueError('No PV specified')
+        pv_cache = get_signal_cache()
```

Comment:
Shouldn't this just use `cache.get_pv_data()`?

Reply:
Hah you are right d'oh